pandas/tools/merge.py

```diff
@@ -199,10 +206,50 @@ def get_result(self):
         typ = self.left._constructor
         result = typ(result_data).__finalize__(self, method='merge')
 
+        if self.indicator:
+            result = self._indicator_post_merge(result)
+
         self._maybe_add_join_keys(result, left_indexer, right_indexer)
 
         return result
 
+    def _indicator_pre_merge(self, left, right):
+
+        if not isinstance(self.indicator, bool) and not isinstance(self.indicator, 
+                                                               compat.string_types):
+            raise ValueError('indicator option can only accept boolean or string arguments')
+        
+        columns = left.columns.values.tolist() + right.columns.values.tolist()  
+
+        for i in ['_left_indicator', '_right_indicator']:
+            if i in columns:
+                raise ValueError("Cannot use `indicator=True` option when data contains a column named {}".format(i))
+        if self.indicator_name in columns:
+            raise ValueError("Cannot use name of an existing column for indicator column")
+
+        left = left.copy()
+        right = right.copy()
+
+        left['_left_indicator'] = 1  
+        left['_left_indicator'] = left['_left_indicator'].astype('int8')  
+        
+        right['_right_indicator'] = 2     
+        right['_right_indicator'] = right['_right_indicator'].astype('int8') 
+        
+        return left, right
+
+    def _indicator_post_merge(self, result):
+
+        result['_left_indicator'] = result['_left_indicator'].fillna(0)
+        result['_right_indicator'] = result['_right_indicator'].fillna(0)
+
+        result[self.indicator_name] = Categorical((result['_left_indicator'] + result['_right_indicator']), categories=[1,2,3])
```

Comment:
The flow is:
    - Add a column to each input data frame. To left, make column all `1`s. To right, make column all `2`s. 
    - After merge, fill missing to zeros, then sum the two columns. (Otherwise we'd have lots of `1` + np.nan if an observation was only in `left`.)
    - if only in `left`, sum is 1. If only in `right`, sum is `2`. If in both, sum is `3`.